{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FUNCTIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import itertools\n",
    "\n",
    "\n",
    "def translate_sequence(sequence, codontab):\n",
    "    protein_seq = \"\"\n",
    "\n",
    "    for i in range(0, len(sequence), 3):\n",
    "        codon = sequence[i:i+3]\n",
    "        if codon in codontab:\n",
    "            protein_seq += codontab[codon]\n",
    "\n",
    "    return protein_seq\n",
    "\n",
    "def get_codon_pairs(sequence):\n",
    "        codon_pairs = []\n",
    "        start_positions = []\n",
    "\n",
    "        for i, amino_acid in enumerate(sequence):\n",
    "            if amino_acid == 'M':\n",
    "                start_positions.append(i)\n",
    "            elif amino_acid == '*':\n",
    "                if start_positions:\n",
    "                    for m_pos in start_positions:\n",
    "                        codon_pairs.append((m_pos, i))\n",
    "                    start_positions = [] \n",
    "        return codon_pairs\n",
    "\n",
    "def find_longest_codon_tuples(codon_pairs):\n",
    "    longest_codon_tuples = []\n",
    "    seen_second_numbers = set()\n",
    "    for tup in codon_pairs:\n",
    "        if tup[1] not in seen_second_numbers:\n",
    "            seen_second_numbers.add(tup[1])\n",
    "            longest_codon_tuples.append(tup)\n",
    "    return longest_codon_tuples\n",
    "\n",
    "def get_sequences(protein_seq, codon_tuples):\n",
    "    sequences = []\n",
    "    for start, end in codon_tuples:\n",
    "        sequence = protein_seq[start:end+1]\n",
    "        sequences.append(sequence)\n",
    "    return sequences\n",
    "\n",
    "def filter_long_sequences(codon_sequences):\n",
    "    return [seq for seq in codon_sequences if len(seq) * 3 > 100]\n",
    "\n",
    "amino_acids = ['A', 'R', 'N', 'D', 'C', 'E', 'Q', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V', '*']\n",
    "def count_codon_dicodon_frequencies(sequences):\n",
    "    codon_freq = defaultdict(int)\n",
    "    dicodon_freq = defaultdict(int)\n",
    "    dicodons = [''.join(pair) for pair in itertools.product(amino_acids, repeat=2)]\n",
    "    codon_freq = {codon: 0 for codon in amino_acids}\n",
    "    dicodon_freq = {dicodon: 0 for dicodon in dicodons}\n",
    "    total_codons = 0\n",
    "    total_dicodons = 0\n",
    "    for sequence in sequences:\n",
    "        total_codons += len(sequence)\n",
    "        total_dicodons += len(sequence) - 1\n",
    "        for codon in sequence:\n",
    "            codon_freq[codon] += 1\n",
    "        for i in range(len(sequence) - 1):\n",
    "            dicodon = sequence[i:i+2]\n",
    "            dicodon_freq[dicodon] += 1\n",
    "            \n",
    "    codon_freq = {codon: freq / total_codons for codon, freq in codon_freq.items()}\n",
    "    dicodon_freq = {dicodon: freq / total_dicodons for dicodon, freq in dicodon_freq.items()}\n",
    "\n",
    "    highest_pair = max(codon_freq, key=codon_freq.get)\n",
    "    print(highest_pair)\n",
    "    highest_pair = max(dicodon_freq, key=dicodon_freq.get)\n",
    "    print(highest_pair)\n",
    "    #print(total_codons)\n",
    "    #print(total_dicodons)\n",
    "    return codon_freq, dicodon_freq\n",
    "\n",
    "def sum_dict_values(codon_dict, dicodon_dict):\n",
    "    codon_total = sum(codon_dict.values())\n",
    "    dicodon_total = sum(dicodon_dict.values())\n",
    "\n",
    "    return codon_total, dicodon_total\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "codontab = {\n",
    "    'TCA': 'S',    # Serina\n",
    "    'TCC': 'S',    # Serina\n",
    "    'TCG': 'S',    # Serina\n",
    "    'TCT': 'S',    # Serina\n",
    "    'TTC': 'F',    # Fenilalanina\n",
    "    'TTT': 'F',    # Fenilalanina\n",
    "    'TTA': 'L',    # Leucina\n",
    "    'TTG': 'L',    # Leucina\n",
    "    'TAC': 'Y',    # Tirosina\n",
    "    'TAT': 'Y',    # Tirosina\n",
    "    'TAA': '*',    # Stop\n",
    "    'TAG': '*',    # Stop\n",
    "    'TGC': 'C',    # Cisteina\n",
    "    'TGT': 'C',    # Cisteina\n",
    "    'TGA': '*',    # Stop\n",
    "    'TGG': 'W',    # Triptofano\n",
    "    'CTA': 'L',    # Leucina\n",
    "    'CTC': 'L',    # Leucina\n",
    "    'CTG': 'L',    # Leucina\n",
    "    'CTT': 'L',    # Leucina\n",
    "    'CCA': 'P',    # Prolina\n",
    "    'CCC': 'P',    # Prolina\n",
    "    'CCG': 'P',    # Prolina\n",
    "    'CCT': 'P',    # Prolina\n",
    "    'CAC': 'H',    # Histidina\n",
    "    'CAT': 'H',    # Histidina\n",
    "    'CAA': 'Q',    # Glutamina\n",
    "    'CAG': 'Q',    # Glutamina\n",
    "    'CGA': 'R',    # Arginina\n",
    "    'CGC': 'R',    # Arginina\n",
    "    'CGG': 'R',    # Arginina\n",
    "    'CGT': 'R',    # Arginina\n",
    "    'ATA': 'I',    # Isoleucina\n",
    "    'ATC': 'I',    # Isoleucina\n",
    "    'ATT': 'I',    # Isoleucina\n",
    "    'ATG': 'M',    # Methionina\n",
    "    'ACA': 'T',    # Treonina\n",
    "    'ACC': 'T',    # Treonina\n",
    "    'ACG': 'T',    # Treonina\n",
    "    'ACT': 'T',    # Treonina\n",
    "    'AAC': 'N',    # Asparagina\n",
    "    'AAT': 'N',    # Asparagina\n",
    "    'AAA': 'K',    # Lisina\n",
    "    'AAG': 'K',    # Lisina\n",
    "    'AGC': 'S',    # Serina\n",
    "    'AGT': 'S',    # Serina\n",
    "    'AGA': 'R',    # Arginina\n",
    "    'AGG': 'R',    # Arginina\n",
    "    'GTA': 'V',    # Valina\n",
    "    'GTC': 'V',    # Valina\n",
    "    'GTG': 'V',    # Valina\n",
    "    'GTT': 'V',    # Valina\n",
    "    'GCA': 'A',    # Alanina\n",
    "    'GCC': 'A',    # Alanina\n",
    "    'GCG': 'A',    # Alanina\n",
    "    'GCT': 'A',    # Alanina\n",
    "    'GAC': 'D',    # Acido Aspartico\n",
    "    'GAT': 'D',    # Acido Aspartico\n",
    "    'GAA': 'E',    # Acido Glutamico\n",
    "    'GAG': 'E',    # Acido Glutamico\n",
    "    'GGA': 'G',    # Glicina\n",
    "    'GGC': 'G',    # Glicina\n",
    "    'GGG': 'G',    # Glicina\n",
    "    'GGT': 'G'     # Glicina\n",
    "}\n",
    "\n",
    "\n",
    "def process_fasta(sequence):\n",
    "    rev_sequence = ''.join(complement[nucleotide] for nucleotide in sequence)\n",
    "    rev_sequence = ''.join(reversed(rev_sequence))\n",
    "\n",
    "    sequence = translate_sequence(sequence, codontab)\n",
    "    rev_sequence = translate_sequence(rev_sequence, codontab)\n",
    "\n",
    "    codon_pairs_pos = get_codon_pairs(sequence)\n",
    "    reverse_codon_pairs_pos = get_codon_pairs(rev_sequence)\n",
    "\n",
    "    longest_codon_pairs_pos = find_longest_codon_tuples(codon_pairs_pos)\n",
    "    longest_reverse_codon_pairs_pos = find_longest_codon_tuples(reverse_codon_pairs_pos)\n",
    "\n",
    "    protein_sequences = get_sequences(sequence, longest_codon_pairs_pos)\n",
    "    reverse_protein_sequences = get_sequences(rev_sequence, longest_reverse_codon_pairs_pos)\n",
    "    \n",
    "    filtered_codon_seq = filter_long_sequences(protein_sequences)\n",
    "    filtered_rev_codon_seq = filter_long_sequences(reverse_protein_sequences)\n",
    "\n",
    "    amino_sequences = filtered_codon_seq + filtered_rev_codon_seq\n",
    "    codon_dict, dicodon_dict = count_codon_dicodon_frequencies(amino_sequences)\n",
    "    codon_vector = list(codon_dict.values())\n",
    "    decodon_vector = list(dicodon_dict.values())\n",
    "    return codon_vector, decodon_vector\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CODE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "def euclidean_distance(vector1, vector2):\n",
    "    vector1 = np.array(vector1)\n",
    "    vector2 = np.array(vector2)\n",
    "    \n",
    "    return np.sqrt(np.sum((vector1 - vector2) ** 2))\n",
    "\n",
    "def calculate_distance_matrix(freq_list):\n",
    "    names = [item[0] for item in freq_list]\n",
    "    vectors = [item[1] for item in freq_list]\n",
    "    \n",
    "    num_vectors = len(vectors)\n",
    "    distance_matrix = np.zeros((num_vectors, num_vectors))\n",
    "    \n",
    "    for i in range(num_vectors):\n",
    "        for j in range(num_vectors):\n",
    "            if i != j:\n",
    "                distance_matrix[i][j] = euclidean_distance(vectors[i], vectors[j])\n",
    "    \n",
    "    distance_matrix = np.round(distance_matrix, 3)\n",
    "    \n",
    "    for i in range(num_vectors):\n",
    "        print(f\"{names[i]} \" + \" \".join(f\"{distance_matrix[i][j]:.3f}\" for j in range(num_vectors)))\n",
    "\n",
    "    return distance_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "codon_freq_list = []\n",
    "dicodon_freq_list = []\n",
    "fasta_directory = r\"C:\\Users\\User\\Desktop\\BioInformatika\\1_lab\\viruses\\data\"\n",
    "for filename in os.listdir(fasta_directory):\n",
    "    if filename.endswith(\".fasta\"):\n",
    "        fasta_file = os.path.join(fasta_directory, filename)\n",
    "        for record in SeqIO.parse(fasta_file, \"fasta\"):  \n",
    "            seq = record.seq\n",
    "            #print(record.name)\n",
    "            codon_vector, dicodon_vector = process_fasta(seq)\n",
    "            codon_freq_list.append((record.name, codon_vector))\n",
    "            dicodon_freq_list.append((record.name, dicodon_vector))\n",
    "\n",
    "print(codon_freq_list)\n",
    "print(len(codon_freq_list))\n",
    "distance_matrix = calculate_distance_matrix(codon_freq_list)\n",
    "print(len(dicodon_freq_list))\n",
    "distance_matrix = calculate_distance_matrix(dicodon_freq_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
